.github/workflows/docker-build.yml

```diff
@@ -11,6 +11,10 @@ env:
   IMAGE_NAME: ${{ github.repository }}
   REGISTRY: ghcr.io
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
```

Comment:
So, one issue with this setup is that it cancels CI on the default branch if a new PR is merged before CI completes. Cancelling CI in a PR is fine because it's in active development. But for `main` we actually want CI to run to completion on every merge, especially for CI with artifacts like this one.

Is there some way to make this cancel CI only for PRs?

Reply:
Good points. With that in mind, this is what I propose:

**Go CI & labels:** Based on the [docs](https://docs.github.com/en/actions/using-jobs/using-concurrency), the suggested way to run on PRs only is `${{ github.head_ref || github.run_id }}`; `head_ref` will only exist on PRs, and if it doesn't exist, concurrency will collate on `run_id`, which is unique for each workflow run, meaning workflows won't be cancelled for non-PR related triggers.

**docker-build:** Barring any billing concerns, I see no harm in letting docker-builds complete and push sha tags to the container store. `${{ github.workflow }}-${{ github.sha }}` means that `docker-build` won't be cancelled unless two runs are triggered on the same git commit. 

I could see this theoretically occurring if a branch is pushed to and then quickly merged into `main`, but we're only tagging by sha anyway, so in any case the tagged artifact will still get created. This could potentially cause problems with #885, if multiple docker-builds trigger on release, targeting different tags, and race or something. An alternative would be to just skip concurrency on `docker-build` entirely and let all builds complete.

Reply:
No billing concerns on docker-build. It's all open source so GH makes it free. I'm in favor of this and the node team can report back and tell us if the new workflow is more convenient.

Reply:
@Wondertan @adlerjohn I believe this thread is resolved. Let me explain:

Github lets us define a concurrency tag format for each workflow. When a workflow gets queued, if it has the same tag as another running workflow, the older workflow gets cancelled (if we have `cancel-in-progress` set). For our purposes, the docs recommend the following tag:

---

`group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}`
* Example tags: `go-ci-feature_branch` if the workflow is from a pull request, or `go-ci-<unique number>` if not.
* Because the tags for a feature branch PR are always the same, new `go-ci`, `docker-build` and `labels` workflows in a branch will cancel any currently running workflows in that branch. If we're NOT in a feature branch (`main`), `head_ref` is not set, and the unique `run_id` will instead ensure that all workflows complete. 

This is essentially the same as what I described in my previous post, but I made `docker-build` use the same cancelling behavior as the other workflows.

Reply:
Ok, thanks for explaining.

Reply:
lgtm